Language/Haskell/GhcMod/Monad/Types.hs

```diff
@@ -129,10 +138,11 @@ data GhcModState = GhcModState {
       gmGhcSession   :: !(Maybe GmGhcSession)
     , gmComponents   :: !(Map ChComponentName (GmComponent GMCResolved (Set ModulePath)))
     , gmCompilerMode :: !CompilerMode
+    , gmMMappedFiles :: Map TargetId (StringBuffer, UTCTime)
```

Comment:
> Ok, I think I kind of understand the idea now. A couple points you
> might have missed though (or where I lack understanding maybe):

> 1. Imports don't generally honor mappings, 

This I don't understand, how would any of this work if an import
didn't honor the mapping??

> and only way to make this
> happen is load all mapped files as targets, which doesn't sound that
> great.

But that's exactly what you have to do to get ghc to accept your
remapped module, is it not? .. I'm so confused.

> I don't think this is a *huge* problem, since, in general, it
> only makes sense to map a couple files that are actually being passed
> as targets, and not whole project.

Well that depends entirely on whay you are doing. If you want to build
a cloud IDE or something based on ghc-mod you'd want to be able to map
the whole project so you don't need to write modules out to the
filesystem maybe.

> 2. Only command where ghc-mod accepts module name is `browse` I
> think, so for other commands `TargetFile` is looked up as `TargetFile`

That's not quite true module names are accepted all over the place.
Most notably also by `check`. Anything that uses `GmlT` can use either
module or file names.

> 3. `guessTarget` favors `TargetFile` greatly: if it can interpret a
> given `String` as a filename, it will.

But only if the file actually exists IIRC from the docs.

> So from where I stand, it seems like concerns would be that `browse`
> and `HomeModuleGraph` don't honor mappings.

> One way to fix this would be to parse module name from source files
> when loading mappings and store those in `Map ModuleName ...`,

I already do that anyways, you just need to add support for remapping.
HMG deals exclusively in fully resolved Module+FilePath units
(`ModulePath`). `resolveModule` implements the FilePath -> ModulePath
mapping and it already does use `fileModuleName` to parse the module
and extract the module name. So your remapping code would have to go
somewhere between `resolveModule` and `updateHomeModuleGraph`.

> and
> then get filenames from HMG when loading targets. But that doesn't
> strike me as the most straightforward option, since,

> in general, we already know filenames when loading mapped files.

We can't just ignore those corner cases though. ghc-mod has to work in
any and all circumstances (within reason).

> [...]

> Scratch that about having both in gmMMappedFiles, that's not very
> feasible. But I still maintain that matching on canonical filenames is
> more robust than matching on module names, since latter is not always
> well-defined.

Let's enumerate ~~all~~ some of the possible cases:

Assuming we have a project with the following module graph:

    A <- B <- D
     \       /
      C <----

With the filesystem layout being the following

    proj.cabal
    `- src
       `- A.hs
        - B.hs
        - C.hs
        - D.hs

Let's look at a few different mappings given the following scenarios:

- The user references module `A` but src/A.hs has a preprocessor error.

    1) module A -> /tmp/A.hs

       That's fine since the user referenced that exact module name
       which we can match on.

    2) src/A.hs -> /tmp/A.hs

       That's fine since when looking for a module called `A` we can use ghc's
       search path algorithm. We will eventually look in src/ since it ought
       to be on the search path. And we'll find a file there matching the
       module name convention therefore that must be the module A.

- The user references the file `src/A.hs` directly but again the
  original has a preprocessor error

   1) module A -> /tmp/A.hs

      This case is more tricky. We have maybe a few possibilities here:

      - We could try to open src/A.hs and parse it's module name but
        as you not this is not always well defined.

      - Apply the search path algorithm in reverse and realize the fact
        that since src/ is on the search path the file called A.hs in this
        directory must be the module A! Otherwise ghc wouldn't find it when
        it's looking for it anyways. Now this would be a bit finicky since
        you have to be careful with the order of directories on the search
        path but I imagine this is feasible.

   2) src/A.hs -> /tmp/A.hs

      This is fine since we can just match on the canonicalized path.

Can you think of any others?